File: 20220810/모범답안/20220810_06.js/FoundMatchedWord.py
def solution(puzzle, word):
    """
    :param puzzle: str[][]
    :param word: str
    :return: boolean
    """

    for x in range(SIZE):
        for y in range(SIZE):
            if move(puzzle, word, x, y, 0): return True

    return False


SIZE = 4
DIRECTIONS = [
    [-1, 0],
    [0, 1],
    [1, 0],
    [0, -1],
]


def move(puzzle, word, x, y, moveIndex):
    if puzzle[x][y] != word[moveIndex]: return False
    if moveIndex == len(word) - 1: return True

    puzzle[x][y] = 'x'

    for [dx, dy] in DIRECTIONS:
        dirX = x + dx
        dirY = y + dy
        if 0 <= dirX and dirX < SIZE and 0 <= dirY and dirY < SIZE:
            if move(puzzle, word, dirX, dirY, moveIndex + 1): return True

    return False

File: 20220810/모범답안/20220810_06.js/FoundMatchedWord.js
/**
 * @param puzzle {string[][]}
 * @param word {string}
 * @returns {boolean}
 */
function solution(puzzle, word) {
  for (let x = 0; x < SIZE; x++) {
    for (let y = 0; y < SIZE; y++) {
      if (move(puzzle, word, x, y, 0)) return true;
    }
  }

  return false;
}

const SIZE = 4;
const DIRECTIONS = [
  [-1, 0],
  [0, 1],
  [1, 0],
  [0, -1],
];

function move(puzzle, word, x, y, moveIndex) {
  if (puzzle[x][y] !== word[moveIndex]) return false;
  if (moveIndex === word.length - 1) return true;

  puzzle[x][y] = 'x';

  for (const [dx, dy] of DIRECTIONS) {
    const dirX = x + dx;
    const dirY = y + dy;
    if (0 <= dirX && dirX < SIZE && 0 <= dirY && dirY < SIZE) {
      if (move(puzzle, word, dirX, dirY, moveIndex + 1)) return true;
    }
  }

  return false;
}

solution